File: app/src/main/java/com/crossbowffs/quotelock/utils/Xlog.java
package com.crossbowffs.quotelock.utils;

import android.util.Log;

import com.crossbowffs.quotelock.BuildConfig;

public final class Xlog {
    private static final String LOG_TAG = BuildConfig.LOG_TAG;
    private static final int LOG_LEVEL = BuildConfig.LOG_LEVEL;
    private static final boolean LOG_TO_XPOSED = BuildConfig.LOG_TO_XPOSED;

    private Xlog() {
    }

    private static void log(int priority, String tag, String message, Object... args) {
        if (priority < LOG_LEVEL) {
            return;
        }

        message = String.format(message, args);
        if (args.length > 0 && args[args.length - 1] instanceof Throwable) {
            Throwable throwable = (Throwable) args[args.length - 1];
            String stacktraceStr = Log.getStackTraceString(throwable);
            message = message + '\n' + stacktraceStr;
        }

        Log.println(priority, LOG_TAG, tag + ": " + message);
        if (LOG_TO_XPOSED) {
            Log.println(priority, "Xposed", LOG_TAG + ": " + tag + ": " + message);
        }
    }

    public static void v(String tag, String message, Object... args) {
        log(Log.VERBOSE, tag, message, args);
    }

    public static void d(String tag, String message, Object... args) {
        log(Log.DEBUG, tag, message, args);
    }

    public static void i(String tag, String message, Object... args) {
        log(Log.INFO, tag, message, args);
    }

    public static void w(String tag, String message, Object... args) {
        log(Log.WARN, tag, message, args);
    }

    public static void e(String tag, String message, Object... args) {
        log(Log.ERROR, tag, message, args);
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/utils/Xlog.kt
package com.crossbowffs.quotelock.utils

import android.util.Log
import com.yubyf.quotelockx.BuildConfig


private const val LOG_TAG = BuildConfig.LOG_TAG
private const val LOG_LEVEL = BuildConfig.LOG_LEVEL
private const val LOG_TO_XPOSED = BuildConfig.LOG_TO_XPOSED

inline fun <reified T> className(): String {
    return T::class.simpleName ?: T::class.java.simpleName
}

object Xlog {
    private fun log(priority: Int, tag: String?, message: String, vararg args: Any) {
        if (priority < LOG_LEVEL) {
            return
        }
        val msg = if (args.isNotEmpty()) {
            String.format(message, *args).let {
                if (args[args.size - 1] is Throwable) {
                    val throwable = args[args.size - 1] as Throwable
                    val stacktraceStr = Log.getStackTraceString(throwable)
                    "$it\n$stacktraceStr"
                } else {
                    it
                }
            }
        } else {
            message
        }
        Log.println(priority, LOG_TAG, "${tag?.let { "$it:" }} $msg")
        if (LOG_TO_XPOSED) {
            Log.println(priority, "Xposed", "$LOG_TAG: ${tag?.let { "$it: " }}$msg")
        }
    }

    fun v(tag: String?, message: String, vararg args: Any) {
        log(Log.VERBOSE, tag, message, *args)
    }

    fun d(tag: String?, message: String, vararg args: Any) {
        log(Log.DEBUG, tag, message, *args)
    }

    fun i(tag: String?, message: String, vararg args: Any) {
        log(Log.INFO, tag, message, *args)
    }

    fun w(tag: String?, message: String, vararg args: Any) {
        log(Log.WARN, tag, message, *args)
    }

    fun e(tag: String?, message: String, vararg args: Any) {
        log(Log.ERROR, tag, message, *args)
    }
}